// @angular-extensions/elements/__ivy_ngcc__/fesm2015/angular-extensions-elements-testing.js
import { Directive, ViewContainerRef, TemplateRef, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-element-testing.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
class LazyElementTestingDirective {
    /**
     * @param {?} vcr
     * @param {?} template
     */
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementTestingDirective.ɵfac = function LazyElementTestingDirective_Factory(t) { return new (t || LazyElementTestingDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
LazyElementTestingDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LazyElementTestingDirective, selectors: [["", "axLazyElement", ""]] });
/** @nocollapse */
LazyElementTestingDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LazyElementTestingDirective, [{
        type: Directive,
        args: [{
                selector: '[axLazyElement]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.TemplateRef }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    LazyElementTestingDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementTestingDirective.prototype.template;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-element-dynamic-testing.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LazyElementDynamicTestingDirective {
    /**
     * @param {?} vcr
     * @param {?} template
     */
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementDynamicTestingDirective.ɵfac = function LazyElementDynamicTestingDirective_Factory(t) { return new (t || LazyElementDynamicTestingDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
LazyElementDynamicTestingDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LazyElementDynamicTestingDirective, selectors: [["", "axLazyElementDynamic", ""]] });
/** @nocollapse */
LazyElementDynamicTestingDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LazyElementDynamicTestingDirective, [{
        type: Directive,
        args: [{
                selector: '[axLazyElementDynamic]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.TemplateRef }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    LazyElementDynamicTestingDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementDynamicTestingDirective.prototype.template;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-elements-testing.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LazyElementsTestingModule {
}
LazyElementsTestingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LazyElementsTestingModule });
LazyElementsTestingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LazyElementsTestingModule_Factory(t) { return new (t || LazyElementsTestingModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LazyElementsTestingModule, { declarations: [LazyElementTestingDirective, LazyElementDynamicTestingDirective], exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LazyElementsTestingModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LazyElementTestingDirective,
                    LazyElementDynamicTestingDirective,
                ],
                exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: angular-extensions-elements-testing.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { LazyElementDynamicTestingDirective, LazyElementTestingDirective, LazyElementsTestingModule };
