// @angular-extensions/elements/bundles/angular-extensions-elements.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@angular-extensions/elements', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory((global['angular-extensions'] = global['angular-extensions'] || {}, global['angular-extensions'].elements = {}), global.ng.core, global.ng.common));
}(this, (function (exports, i0, common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lazy-elements/lazy-elements.tokens.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LAZY_ELEMENT_CONFIGS = new i0.InjectionToken('LAZY_ELEMENT_CONFIGS');
    /** @type {?} */
    var LAZY_ELEMENT_ROOT_OPTIONS = new i0.InjectionToken('LAZY_ELEMENT_ROOT_OPTIONS');
    /** @type {?} */
    var LAZY_ELEMENT_ROOT_GUARD = new i0.InjectionToken('LAZY_ELEMENT_ROOT_GUARD');
    /** @type {?} */
    var LAZY_ELEMENTS_REGISTRY = new i0.InjectionToken('LAZY_ELEMENTS_REGISTRY', {
        providedIn: 'root',
        factory: ( /**
         * @return {?}
         */function () { return new Map(); }),
    });
    /**
     * @record
     */
    function LazyElementsRegistry() { }
    if (false) {
        /** @type {?} */
        LazyElementsRegistry.prototype.get;
        /** @type {?} */
        LazyElementsRegistry.prototype.set;
        /** @type {?} */
        LazyElementsRegistry.prototype.has;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lazy-elements/lazy-elements-loader.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LOG_PREFIX = '@angular-extensions/elements';
    /**
     * @record
     */
    function HooksConfig() { }
    if (false) {
        /** @type {?|undefined} */
        HooksConfig.prototype.beforeLoad;
        /** @type {?|undefined} */
        HooksConfig.prototype.afterLoad;
    }
    /**
     * @record
     */
    function ElementConfig() { }
    if (false) {
        /** @type {?} */
        ElementConfig.prototype.tag;
        /** @type {?} */
        ElementConfig.prototype.url;
        /** @type {?|undefined} */
        ElementConfig.prototype.isModule;
        /** @type {?|undefined} */
        ElementConfig.prototype.loadingComponent;
        /** @type {?|undefined} */
        ElementConfig.prototype.errorComponent;
        /** @type {?|undefined} */
        ElementConfig.prototype.preload;
        /** @type {?|undefined} */
        ElementConfig.prototype.hooks;
        /** @type {?|undefined} */
        ElementConfig.prototype.isAdded;
    }
    var LazyElementsLoaderService = /** @class */ (function () {
        /**
         * @param {?} registry
         * @param {?} options
         */
        function LazyElementsLoaderService(registry, options) {
            this.registry = registry;
            this.options = options;
            this.configs = [];
            if (!options) {
                this.options = {};
            }
        }
        /**
         * @param {?} newConfigs
         * @return {?}
         */
        LazyElementsLoaderService.prototype.addConfigs = function (newConfigs) {
            var _this = this;
            newConfigs.forEach(( /**
             * @param {?} newConfig
             * @return {?}
             */function (newConfig) {
                /** @type {?} */
                var existingConfig = _this.getElementConfig(newConfig.tag);
                if (existingConfig) {
                    console.warn(LOG_PREFIX + " - ElementConfig for tag '" + newConfig.tag + "' was previously added, it will not be added multiple times, continue...");
                }
                else {
                    newConfig.isAdded = true;
                    _this.configs.push(newConfig);
                    /** @type {?} */
                    var shouldPreload = newConfig.preload !== undefined
                        ? newConfig.preload
                        : _this.options.preload;
                    if (shouldPreload) {
                        _this.loadElement(newConfig.url, newConfig.tag, newConfig.isModule, newConfig.hooks);
                    }
                }
            }));
        };
        /**
         * @param {?} tag
         * @return {?}
         */
        LazyElementsLoaderService.prototype.getElementConfig = function (tag) {
            return this.configs.find(( /**
             * @param {?} config
             * @return {?}
             */function (config) { return config.tag === tag; }));
        };
        /**
         * @param {?=} tags
         * @return {?}
         */
        LazyElementsLoaderService.prototype.preload = function (tags) {
            var _this = this;
            /** @type {?} */
            var configs = this.configs;
            if (tags) {
                configs = this.configs.filter(( /**
                 * @param {?} config
                 * @return {?}
                 */function (config) { return tags.includes(config.tag); }));
            }
            configs.forEach(( /**
             * @param {?} config
             * @return {?}
             */function (config) { return _this.loadElement(config.url, config.tag, config.isModule, config.hooks); }));
        };
        /**
         * @param {?} url
         * @param {?} tag
         * @param {?=} isModule
         * @param {?=} hooksConfig
         * @return {?}
         */
        LazyElementsLoaderService.prototype.loadElement = function (url, tag, isModule, hooksConfig) {
            var _this = this;
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
            /** @type {?} */
            var config = this.getElementConfig(tag);
            if (!url) {
                if (!config || !config.url) {
                    throw new Error(LOG_PREFIX + " - url for <" + tag + "> not found");
                }
                url = config.url;
            }
            if (isModule === undefined) {
                isModule =
                    config && config.isModule !== undefined
                        ? config.isModule
                        : this.options.isModule;
            }
            if (!tag) {
                throw new Error(LOG_PREFIX + " - tag for '" + url + "' not found, the *axLazyElement has to be used on HTML element");
            }
            if (!this.hasElement(url)) {
                /** @type {?} */
                var notifier_1 = this.addElement(url);
                /** @type {?} */
                var script_1 = ( /** @type {?} */(document.createElement('script')));
                /** @type {?} */
                var beforeLoadHook = (_c = (_a = hooksConfig === null || hooksConfig === void 0 ? void 0 : hooksConfig.beforeLoad) !== null && _a !== void 0 ? _a : (_b = config === null || config === void 0 ? void 0 : config.hooks) === null || _b === void 0 ? void 0 : _b.beforeLoad) !== null && _c !== void 0 ? _c : (_e = (_d = this.options) === null || _d === void 0 ? void 0 : _d.hooks) === null || _e === void 0 ? void 0 : _e.beforeLoad;
                /** @type {?} */
                var afterLoadHook_1 = (_h = (_f = hooksConfig === null || hooksConfig === void 0 ? void 0 : hooksConfig.afterLoad) !== null && _f !== void 0 ? _f : (_g = config === null || config === void 0 ? void 0 : config.hooks) === null || _g === void 0 ? void 0 : _g.afterLoad) !== null && _h !== void 0 ? _h : (_k = (_j = this.options) === null || _j === void 0 ? void 0 : _j.hooks) === null || _k === void 0 ? void 0 : _k.afterLoad;
                if (isModule) {
                    script_1.type = 'module';
                }
                script_1.src = url;
                script_1.onload = ( /**
                 * @return {?}
                 */function () {
                    if (afterLoadHook_1) {
                        _this.handleHook(afterLoadHook_1, tag)
                            .then(notifier_1.resolve)
                            .catch(notifier_1.reject);
                    }
                    else {
                        notifier_1.resolve();
                    }
                });
                script_1.onerror = notifier_1.reject;
                if (beforeLoadHook) {
                    this.handleHook(beforeLoadHook, tag)
                        .then(( /**
                 * @return {?}
                 */function () { return document.body.appendChild(script_1); }))
                        .catch(notifier_1.reject);
                }
                else {
                    document.body.appendChild(script_1);
                }
            }
            return this.registry.get(this.stripUrlProtocol(url));
        };
        /**
         * @private
         * @param {?} url
         * @return {?}
         */
        LazyElementsLoaderService.prototype.addElement = function (url) {
            /** @type {?} */
            var notifier;
            this.registry.set(this.stripUrlProtocol(url), new Promise(( /**
             * @param {?} resolve
             * @param {?} reject
             * @return {?}
             */function (resolve, reject) { return (notifier = { resolve: resolve, reject: reject }); })));
            return notifier;
        };
        /**
         * @private
         * @param {?} url
         * @return {?}
         */
        LazyElementsLoaderService.prototype.hasElement = function (url) {
            return this.registry.has(this.stripUrlProtocol(url));
        };
        /**
         * @private
         * @param {?} url
         * @return {?}
         */
        LazyElementsLoaderService.prototype.stripUrlProtocol = function (url) {
            return url.replace(/https?:\/\//, '');
        };
        /**
         * @private
         * @template T
         * @param {?} obj
         * @return {?}
         */
        LazyElementsLoaderService.prototype.isPromise = function (obj) {
            var _a;
            return typeof ((_a = (( /** @type {?} */(obj)))) === null || _a === void 0 ? void 0 : _a.then) === 'function';
        };
        /**
         * @private
         * @param {?} hook
         * @param {?} tag
         * @return {?}
         */
        LazyElementsLoaderService.prototype.handleHook = function (hook, tag) {
            try {
                /** @type {?} */
                var result = hook(tag);
                if (this.isPromise(result)) {
                    return result;
                }
                else {
                    return Promise.resolve();
                }
            }
            catch (err) {
                return Promise.reject(err);
            }
        };
        return LazyElementsLoaderService;
    }());
    LazyElementsLoaderService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    LazyElementsLoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: i0.Inject, args: [LAZY_ELEMENTS_REGISTRY,] }] },
        { type: undefined, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [LAZY_ELEMENT_ROOT_OPTIONS,] }] }
    ]; };
    /** @nocollapse */ LazyElementsLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function LazyElementsLoaderService_Factory() { return new LazyElementsLoaderService(i0.ɵɵinject(LAZY_ELEMENTS_REGISTRY), i0.ɵɵinject(LAZY_ELEMENT_ROOT_OPTIONS, 8)); }, token: LazyElementsLoaderService, providedIn: "root" });
    if (false) {
        /** @type {?} */
        LazyElementsLoaderService.prototype.configs;
        /**
         * @type {?}
         * @private
         */
        LazyElementsLoaderService.prototype.registry;
        /** @type {?} */
        LazyElementsLoaderService.prototype.options;
    }
    /**
     * @record
     */
    function Notifier() { }
    if (false) {
        /** @type {?} */
        Notifier.prototype.resolve;
        /** @type {?} */
        Notifier.prototype.reject;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lazy-elements/lazy-element/lazy-element.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LOG_PREFIX$1 = '@angular-extensions/elements';
    var LazyElementDirective = /** @class */ (function () {
        /**
         * @param {?} vcr
         * @param {?} template
         * @param {?} elementsLoaderService
         * @param {?} cfr
         * @param {?} cdr
         */
        function LazyElementDirective(vcr, template, elementsLoaderService, cfr, cdr) {
            this.vcr = vcr;
            this.template = template;
            this.elementsLoaderService = elementsLoaderService;
            this.cfr = cfr;
            this.cdr = cdr;
            // tslint:disable-line:no-input-rename
            this.viewRef = null;
        }
        /**
         * @return {?}
         */
        LazyElementDirective.prototype.ngOnInit = function () {
            var _this = this;
            /** @type {?} */
            var tpl = ( /** @type {?} */(this.template));
            /** @type {?} */
            var elementTag = tpl._declarationTContainer
                ? tpl._declarationTContainer.tagName
                : tpl._def.element.template.nodes[0].element.name;
            /** @type {?} */
            var elementConfig = this.elementsLoaderService.getElementConfig(elementTag) ||
                (( /** @type {?} */({})));
            /** @type {?} */
            var options = this.elementsLoaderService.options;
            /** @type {?} */
            var loadingComponent = elementConfig.loadingComponent || options.loadingComponent;
            if (this.loadingTemplateRef) {
                this.vcr.createEmbeddedView(this.loadingTemplateRef);
            }
            else if (loadingComponent) {
                /** @type {?} */
                var factory = this.cfr.resolveComponentFactory(loadingComponent);
                this.vcr.createComponent(factory);
            }
            this.elementsLoaderService
                .loadElement(this.url, elementTag, this.isModule, elementConfig === null || elementConfig === void 0 ? void 0 : elementConfig.hooks)
                .then(( /**
         * @return {?}
         */function () { return customElements.whenDefined(elementTag); }))
                .then(( /**
         * @return {?}
         */function () {
                _this.vcr.clear();
                _this.viewRef = _this.vcr.createEmbeddedView(_this.template);
                _this.cdr.markForCheck();
            }))
                .catch(( /**
         * @return {?}
         */function () {
                _this.vcr.clear();
                /** @type {?} */
                var errorComponent = elementConfig.errorComponent || options.errorComponent;
                if (_this.errorTemplateRef) {
                    _this.vcr.createEmbeddedView(_this.errorTemplateRef);
                    _this.cdr.markForCheck();
                }
                else if (errorComponent) {
                    /** @type {?} */
                    var factory = _this.cfr.resolveComponentFactory(errorComponent);
                    _this.vcr.createComponent(factory);
                    _this.cdr.markForCheck();
                }
                else {
                    console.error(LOG_PREFIX$1 + " - Loading of element <" + elementTag + "> failed, please provide <ng-template #error>Loading failed...</ng-template> and reference it in *axLazyElement=\"errorTemplate: error\" to display customized error message in place of element");
                }
            }));
        };
        /**
         * @return {?}
         */
        LazyElementDirective.prototype.destroyEmbeddedView = function () {
            if (this.viewRef && !this.viewRef.destroyed) {
                this.viewRef.detach();
                this.viewRef.destroy();
                this.viewRef = null;
            }
        };
        return LazyElementDirective;
    }());
    LazyElementDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[axLazyElement]',
                },] }
    ];
    /** @nocollapse */
    LazyElementDirective.ctorParameters = function () { return [
        { type: i0.ViewContainerRef },
        { type: i0.TemplateRef },
        { type: LazyElementsLoaderService },
        { type: i0.ComponentFactoryResolver },
        { type: i0.ChangeDetectorRef }
    ]; };
    LazyElementDirective.propDecorators = {
        url: [{ type: i0.Input, args: ['axLazyElement',] }],
        loadingTemplateRef: [{ type: i0.Input, args: ['axLazyElementLoadingTemplate',] }],
        errorTemplateRef: [{ type: i0.Input, args: ['axLazyElementErrorTemplate',] }],
        isModule: [{ type: i0.Input, args: ['axLazyElementModule',] }]
    };
    if (false) {
        /** @type {?} */
        LazyElementDirective.prototype.url;
        /** @type {?} */
        LazyElementDirective.prototype.loadingTemplateRef;
        /** @type {?} */
        LazyElementDirective.prototype.errorTemplateRef;
        /** @type {?} */
        LazyElementDirective.prototype.isModule;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.viewRef;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.vcr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.template;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.elementsLoaderService;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.cfr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDirective.prototype.cdr;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lazy-elements/lazy-element-dynamic/lazy-element-dynamic.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LOG_PREFIX$2 = '@angular-extensions/elements';
    /**
     * \@dynamic
     */
    var LazyElementDynamicDirective = /** @class */ (function () {
        /**
         * @param {?} document
         * @param {?} renderer
         * @param {?} vcr
         * @param {?} cfr
         * @param {?} cdr
         * @param {?} template
         * @param {?} elementsLoaderService
         */
        function LazyElementDynamicDirective(document, renderer, vcr, cfr, cdr, template, elementsLoaderService) {
            this.document = document;
            this.renderer = renderer;
            this.vcr = vcr;
            this.cfr = cfr;
            this.cdr = cdr;
            this.template = template;
            this.elementsLoaderService = elementsLoaderService;
            // tslint:disable-line:no-input-rename
            this.viewRef = null;
        }
        /**
         * @return {?}
         */
        LazyElementDynamicDirective.prototype.ngOnInit = function () {
            var _this = this;
            if (!this.tag || this.tag.length === 0 || !this.tag.includes('-')) {
                throw new Error(LOG_PREFIX$2 + " - Valid tag has to be specified when using *axLazyElementDynamic directive (use *axLazyElementDynamic=\"'some-tag'\"), got: \"" + this.tag + "\"");
            }
            /** @type {?} */
            var elementConfig = this.elementsLoaderService.getElementConfig(this.tag) ||
                (( /** @type {?} */({})));
            /** @type {?} */
            var options = this.elementsLoaderService.options;
            /** @type {?} */
            var loadingComponent = elementConfig.loadingComponent || options.loadingComponent;
            if (this.loadingTemplateRef) {
                this.vcr.createEmbeddedView(this.loadingTemplateRef);
            }
            else if (loadingComponent) {
                /** @type {?} */
                var factory = this.cfr.resolveComponentFactory(loadingComponent);
                this.vcr.createComponent(factory);
            }
            this.elementsLoaderService
                .loadElement(this.url, this.tag, this.isModule, elementConfig === null || elementConfig === void 0 ? void 0 : elementConfig.hooks)
                .then(( /**
         * @return {?}
         */function () { return customElements.whenDefined(_this.tag); }))
                .then(( /**
         * @return {?}
         */function () {
                _this.vcr.clear();
                /** @type {?} */
                var originalCreateElement = _this.renderer.createElement;
                _this.renderer.createElement = ( /**
                 * @param {?} name
                 * @param {?} namespace
                 * @return {?}
                 */function (name, namespace) {
                    if (name === 'ax-lazy-element') {
                        name = _this.tag;
                    }
                    return _this.document.createElement(name);
                });
                _this.viewRef = _this.vcr.createEmbeddedView(_this.template);
                _this.renderer.createElement = originalCreateElement;
                _this.cdr.markForCheck();
            }))
                .catch(( /**
         * @param {?} error
         * @return {?}
         */function (error) {
                /** @type {?} */
                var errorComponent = elementConfig.errorComponent || options.errorComponent;
                _this.vcr.clear();
                if (_this.errorTemplateRef) {
                    _this.vcr.createEmbeddedView(_this.errorTemplateRef);
                    _this.cdr.markForCheck();
                }
                else if (errorComponent) {
                    /** @type {?} */
                    var factory = _this.cfr.resolveComponentFactory(errorComponent);
                    _this.vcr.createComponent(factory);
                    _this.cdr.markForCheck();
                }
                else {
                    console.error(LOG_PREFIX$2 + " - Loading of element <" + _this.tag + "> failed, please provide <ng-template #error>Loading failed...</ng-template> and reference it in *axLazyElementDynamic=\"errorTemplate: error\" to display customized error message in place of element\n\n", error);
                }
            }));
        };
        /**
         * @return {?}
         */
        LazyElementDynamicDirective.prototype.destroyEmbeddedView = function () {
            if (this.viewRef && !this.viewRef.destroyed) {
                this.viewRef.detach();
                this.viewRef.destroy();
                this.viewRef = null;
            }
        };
        return LazyElementDynamicDirective;
    }());
    LazyElementDynamicDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[axLazyElementDynamic]',
                },] }
    ];
    /** @nocollapse */
    LazyElementDynamicDirective.ctorParameters = function () { return [
        { type: Document, decorators: [{ type: i0.Inject, args: [common.DOCUMENT,] }] },
        { type: i0.Renderer2 },
        { type: i0.ViewContainerRef },
        { type: i0.ComponentFactoryResolver },
        { type: i0.ChangeDetectorRef },
        { type: i0.TemplateRef },
        { type: LazyElementsLoaderService }
    ]; };
    LazyElementDynamicDirective.propDecorators = {
        tag: [{ type: i0.Input, args: ['axLazyElementDynamic',] }],
        url: [{ type: i0.Input, args: ['axLazyElementDynamicUrl',] }],
        loadingTemplateRef: [{ type: i0.Input, args: ['axLazyElementDynamicLoadingTemplate',] }],
        errorTemplateRef: [{ type: i0.Input, args: ['axLazyElementDynamicErrorTemplate',] }],
        isModule: [{ type: i0.Input, args: ['axLazyElementDynamicModule',] }]
    };
    if (false) {
        /** @type {?} */
        LazyElementDynamicDirective.prototype.tag;
        /** @type {?} */
        LazyElementDynamicDirective.prototype.url;
        /** @type {?} */
        LazyElementDynamicDirective.prototype.loadingTemplateRef;
        /** @type {?} */
        LazyElementDynamicDirective.prototype.errorTemplateRef;
        /** @type {?} */
        LazyElementDynamicDirective.prototype.isModule;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.viewRef;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.document;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.renderer;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.vcr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.cfr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.cdr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.template;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicDirective.prototype.elementsLoaderService;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lazy-elements/lazy-elements.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @param {?} rootOptions
     * @return {?}
     */
    function createLazyElementRootGuard(rootOptions) {
        if (rootOptions) {
            throw new TypeError("LazyElementsModule.forRoot() called twice. Feature modules should use LazyElementsModule.forFeature() instead.");
        }
        return 'LazyElementsModule.forRoot() multiple execution guard';
    }
    var LazyElementsModule = /** @class */ (function () {
        /**
         * @param {?} lazyElementsLoaderService
         * @param {?} elementConfigsMultiProvider
         * @param {?} guard
         */
        function LazyElementsModule(lazyElementsLoaderService, elementConfigsMultiProvider, guard) {
            if (elementConfigsMultiProvider && elementConfigsMultiProvider.length) {
                elementConfigsMultiProvider
                    .filter(( /**
             * @param {?} configs
             * @return {?}
             */function (configs) { return configs.some(( /**
                 * @param {?} config
                 * @return {?}
                 */function (config) { return !config.isAdded; })); }))
                    .forEach(( /**
             * @param {?} configs
             * @return {?}
             */function (configs) { return lazyElementsLoaderService.addConfigs(configs); }));
            }
        }
        /**
         * @param {?} options
         * @return {?}
         */
        LazyElementsModule.forRoot = function (options) {
            return {
                ngModule: LazyElementsModule,
                providers: [
                    {
                        provide: LAZY_ELEMENT_CONFIGS,
                        useValue: options && options.elementConfigs ? options.elementConfigs : [],
                        multi: true,
                    },
                    {
                        provide: LAZY_ELEMENT_ROOT_OPTIONS,
                        useValue: options.rootOptions ? options.rootOptions : {},
                    },
                    {
                        provide: LAZY_ELEMENT_ROOT_GUARD,
                        useFactory: createLazyElementRootGuard,
                        deps: [[LAZY_ELEMENT_ROOT_OPTIONS, new i0.Optional(), new i0.SkipSelf()]],
                    },
                ],
            };
        };
        /**
         * @param {?} options
         * @return {?}
         */
        LazyElementsModule.forFeature = function (options) {
            return {
                ngModule: LazyElementsModule,
                providers: [
                    {
                        provide: LAZY_ELEMENT_CONFIGS,
                        useValue: options && options.elementConfigs ? options.elementConfigs : [],
                        multi: true,
                    },
                ],
            };
        };
        return LazyElementsModule;
    }());
    LazyElementsModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [LazyElementDirective, LazyElementDynamicDirective],
                    imports: [common.CommonModule],
                    exports: [LazyElementDirective, LazyElementDynamicDirective],
                    providers: [],
                },] }
    ];
    /** @nocollapse */
    LazyElementsModule.ctorParameters = function () { return [
        { type: LazyElementsLoaderService },
        { type: Array, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [LAZY_ELEMENT_CONFIGS,] }] },
        { type: undefined, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [LAZY_ELEMENT_ROOT_GUARD,] }] }
    ]; };
    /**
     * @record
     */
    function LazyElementModuleOptions() { }
    if (false) {
        /** @type {?|undefined} */
        LazyElementModuleOptions.prototype.elementConfigs;
    }
    /**
     * @record
     */
    function LazyElementModuleRootOptions() { }
    if (false) {
        /** @type {?|undefined} */
        LazyElementModuleRootOptions.prototype.elementConfigs;
        /** @type {?|undefined} */
        LazyElementModuleRootOptions.prototype.rootOptions;
    }
    /**
     * @record
     */
    function LazyElementRootOptions() { }
    if (false) {
        /** @type {?|undefined} */
        LazyElementRootOptions.prototype.loadingComponent;
        /** @type {?|undefined} */
        LazyElementRootOptions.prototype.errorComponent;
        /** @type {?|undefined} */
        LazyElementRootOptions.prototype.isModule;
        /** @type {?|undefined} */
        LazyElementRootOptions.prototype.preload;
        /** @type {?|undefined} */
        LazyElementRootOptions.prototype.hooks;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: public_api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: angular-extensions-elements.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.LAZY_ELEMENTS_REGISTRY = LAZY_ELEMENTS_REGISTRY;
    exports.LAZY_ELEMENT_CONFIGS = LAZY_ELEMENT_CONFIGS;
    exports.LAZY_ELEMENT_ROOT_GUARD = LAZY_ELEMENT_ROOT_GUARD;
    exports.LAZY_ELEMENT_ROOT_OPTIONS = LAZY_ELEMENT_ROOT_OPTIONS;
    exports.LazyElementDirective = LazyElementDirective;
    exports.LazyElementDynamicDirective = LazyElementDynamicDirective;
    exports.LazyElementsLoaderService = LazyElementsLoaderService;
    exports.LazyElementsModule = LazyElementsModule;
    exports.createLazyElementRootGuard = createLazyElementRootGuard;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
